File: 2435-PathsinMatrixWhoseSumIsDivisiblebyK/2435-PathsinMatrixWhoseSumIsDivisiblebyK.py
# Last updated: 25/9/2025, 12:21:32 am
class Solution:
    def numberOfPaths(self, grid: List[List[int]], k: int) -> int:
        m = len(grid)
        n = len(grid[0])

        mod = 10**9 + 7

        dp = [[[-1] * k for _ in range(n)] for _ in range(m)]

        def rec(i, j, rem):
            rem = (rem + grid[i][j]) % k
            
            if i == m-1 and j == n-1:
                return 1 if rem == 0 else 0
            
            if dp[i][j][rem] != -1:
                return dp[i][j][rem]

            ans = 0
            if i + 1 < m:
                ans += rec(i + 1, j, rem)
            if j + 1 < n:
                ans += rec(i, j + 1, rem)

            ans %= mod
            dp[i][j][rem] = ans
            return ans

        return rec(0, 0, 0)

File: 2435-PathsinMatrixWhoseSumIsDivisiblebyK/2435-PathsinMatrixWhoseSumIsDivisiblebyK.cpp
// Last updated: 26/11/2025, 6:50:20 am
#include <algorithm>
#include <iostream>
#include <vector>
#include <cmath>
#include <cstring>
#include <unordered_map>
#include <unordered_set>

using namespace std;

const int mod = 1000000007;

class Solution {
public:
    int numberOfPaths(vector<vector<int>>& grid, int k) {
        const int m = grid.size();
        const int n = grid[0].size();

        int dp[m + 1][n + 1][k];
        memset(dp, 0, sizeof(dp));
        dp[0][0][0] = 1;
        for (int i=0; i<m; i++) {
            for (int j=0; j<n; j++) {
                for (int rem=0; rem<k; rem++) {
                    int new_rem = (grid[i][j] + rem) % k;
                    dp[i + 1][j][new_rem] += dp[i][j][rem];
                    dp[i + 1][j][new_rem] %= mod;
                    dp[i][j + 1][new_rem] += dp[i][j][rem];
                    dp[i][j + 1][new_rem] %= mod;
                }
            }
        }

        return dp[m][n-1][0];
    }
};